subsys/serialization/transport/rp_ll.c

```diff
@@ -0,0 +1,335 @@
+
+#define VRING_COUNT         2
+#define VRING_TX_ADDRESS    (SHM_START_ADDR + SHM_SIZE - 0x400)
+#define VRING_RX_ADDRESS    (VRING_TX_ADDRESS - 0x400)
+#define VRING_ALIGNMENT     4
+#define VRING_SIZE          16
+
+#define VDEV_STATUS_ADDR    DT_IPC_SHM_BASE_ADDRESS
+
+
+/* Handlers for TX and RX channels */
+/* TX handler */
+static struct device *ipm_tx_handle;
+
+/* RX handler */
+static struct device *ipm_rx_handle;
+
+static struct rpmsg_virtio_device rvdev;
+struct rpmsg_device *rdev;
+
+static metal_phys_addr_t shm_physmap[] = { SHM_START_ADDR };
+static struct metal_device shm_device = {
+	.name = SHM_DEVICE_NAME,
+	.bus = NULL,
+	.num_regions = 1,
+	.regions = {
+		{
+			.virt = (void *)SHM_START_ADDR,
+			.physmap = shm_physmap,
+			.size = SHM_SIZE,
+			.page_shift = 0xffffffff,
+			.page_mask = 0xffffffff,
+			.mem_flags = 0,
+			.ops = { NULL },
+		},
+	},
+	.node = { NULL },
+	.irq_num = 0,
+	.irq_info = NULL
+};
+
+static struct virtqueue *vq[2];
+
+/* Thread properties */
+static K_THREAD_STACK_DEFINE(rx_thread_stack, STACKSIZE);
+static struct k_thread rx_thread_data;
+
+
+static unsigned char virtio_get_status(struct virtio_device *vdev)
+{
+	return IS_ENABLED(CONFIG_RPMSG_MASTER) ?
+		VIRTIO_CONFIG_STATUS_DRIVER_OK : sys_read8(VDEV_STATUS_ADDR);
+}
+
+static u32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return 0;
+}
+
+void virtio_set_status(struct virtio_device *vdev, unsigned char status)
+{
+	sys_write8(status, VDEV_STATUS_ADDR);
+}
+
+static void virtio_set_features(struct virtio_device *vdev, u32_t features)
+{
+	/* No need for implementation */
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	int status;
+
+	status = ipm_send(ipm_tx_handle, 0, 0, NULL, 0);
+	if (status != 0) {
+		LOG_ERR("Failed to notify: %d", status);
+	}
+}
+
+const struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.get_features = virtio_get_features,
+	.set_status = virtio_set_status,
+	.set_features = virtio_set_features,
+	.notify = virtio_notify,
+};
+
+/* Callback launch right after some data has arrived. */
+static void ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	k_sem_give(&ipm_event_sem);
+}
+
+/* Callback launch right after virtqueue_notification from rx_thread. */
+static int endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
+		       u32_t src, void *priv)
+{
+	struct rp_ll_endpoint *my_ep = metal_container_of(ept,
+		struct rp_ll_endpoint, rpmsg_ep);
+
+	if (len == 0) {
+		if (!(my_ep->flags & EP_FLAG_HANSHAKE_DONE)) {
+			rpmsg_send(ept, (u8_t *)"", 0);
+			my_ep->flags |= EP_FLAG_HANSHAKE_DONE;
+			LOG_INF("Handshake done");
+			my_ep->callback(my_ep, RP_LL_EVENT_CONNECTED, NULL, 0);
+		}
+		return RPMSG_SUCCESS;
+	}
+
+	my_ep->callback(my_ep, RP_LL_EVENT_DATA, data, len);
+
+	return RPMSG_SUCCESS;
+}
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+static void rx_thread(void *p1, void *p2, void *p3)
```

Comment:
This thread is indented to give fast notification to the endpoint about new data, so it should have high priority and pretty fast response. Using system's work queue does not guarantee that.

Reply:
Then use a user workqueue?

Reply:
User workqueue will work, but why add more execution time and memory overhead? Workqueue is actually a thread + queue containing work items. Single thread instead of workqueue will not have overhead related to handling a queue.

Reply:
> why add more execution time and memory overhead?

To remove code duplication.

The additional RAM overhead for a work queue and a work item appears to be 8 bytes (172 vs 180 bytes, excluding stack). There's not really much additional processing done inside either, surely this can't be worth reimplementing a kernel component for?

Reply:
I made a test implementation to compare results. You can see details below. In the example without workqueue before, flash size change is significant and the other things are not. In example that has workqueue already used there is no significant change in any of the metrics. I am preferring to keep current implementation because:
1. Smaller flash size in the first example,
2. This file will be deleted sooner or later.

I have a ready implementation, so if you strongly insist to workqueue I can push it to this PR. 

Thread:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       23808 B         1 MB      2.27%
            SRAM:       11200 B       448 KB      2.44%
Stack usage: 32
Lines of significant code: 15
```
Workqueue:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       23940 B         1 MB      2.28%
            SRAM:       11208 B       448 KB      2.44%
Stack usage: 24
Lines of significant code: 9
```
In total:
+132 byte of flash
+8 bytes of RAM
-8 bytes of stack
-6 lines of code

You can notice significant flash size increase +132 after switching to workqueue. It was because this example was not using workqueue before. Example that use workqueue before:

Dedicated thread:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       24028 B         1 MB      2.29%
            SRAM:       12360 B       448 KB      2.69%
Stack usage: 32
Lines of significant code: 15
```
Dedicated workqueue:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       24020 B         1 MB      2.29%
            SRAM:       12368 B       448 KB      2.70%
Stack usage: 24
Lines of significant code: 9
```
In total:
-8 byte of flash
+8 bytes of RAM
-8 bytes of stack
-6 lines of code

Reply:
The cost of workqueue will be amortized for every user.

Imagine if every user of workqueue made this same argument, then they would all decide to implement their own queue and we would have x implementations, with a total flash size that exceeds workqueue.

Reply:
Going with the kernel module would reduce the amount of code that needs to be maintained and tested.
Unless the optimization is really needed I think we should consider those aspects first. 132 bytes doesn't seem like a significant cost of flash usage.